File: engine/src/reviewer.ts
/** Orchestrates the deterministic + AI hybrid review of a single flow. */
import { parseFlow } from './flowModel.js';
import { evaluatorRegistry } from './rules/evaluators.js';
import type {
  Finding,
  FlowReviewInput,
  FlowReviewResult,
  ReviewOptions,
  ReviewScore,
  ReviewRule,
  Ruleset,
  Severity,
} from './types.js';

const SEVERITY_WEIGHT: Record<Severity, number> = {
  info: 0,
  warning: 1,
  error: 3,
  critical: 6,
};

/** Worst outcome across a set of findings for the same rule. */
function worstStatus(findings: Finding[]): Finding['status'] {
  const order = { not_applicable: 0, pass: 1, warning: 2, fail: 3 } as const;
  let worst: Finding['status'] = 'not_applicable';
  for (const f of findings) if (order[f.status] > order[worst]) worst = f.status;
  return worst;
}

/**
 * Compute a 0–100 compliance score.
 *
 * Each *applicable* rule (one that produced a pass/warning/fail, i.e. was not
 * N/A) has a maximum penalty of `weight × severityWeight(rule.severity)`. Its
 * actual penalty is the full amount if it failed, half if it only warned, and
 * zero if it passed. The score is `1 − Σpenalty / Σmaxpenalty`, so it measures
 * "of the standards that applied to this flow, how many did it meet", weighted
 * by how much each matters — and is unaffected by rules that were N/A.
 */
export function scoreFindings(findings: Finding[], rules: ReviewRule[]): ReviewScore {
  const bySeverity: Record<Severity, number> = { info: 0, warning: 0, error: 0, critical: 0 };
  let passed = 0;
  let failed = 0;
  let warnings = 0;
  let notApplicable = 0;

  for (const f of findings) {
    switch (f.status) {
      case 'pass':
        passed++;
        break;
      case 'warning':
        warnings++;
        bySeverity[f.severity]++;
        break;
      case 'fail':
        failed++;
        bySeverity[f.severity]++;
        break;
      case 'not_applicable':
        notApplicable++;
        break;
    }
  }

  const ruleByCode = new Map(rules.map((r) => [r.code, r]));
  const grouped = new Map<string, Finding[]>();
  for (const f of findings) {
    const list = grouped.get(f.ruleCode) ?? [];
    list.push(f);
    grouped.set(f.ruleCode, list);
  }

  let penalty = 0;
  let maxPenalty = 0;
  for (const [code, group] of grouped) {
    const rule = ruleByCode.get(code);
    const status = worstStatus(group);
    if (status === 'not_applicable') continue;
    const max = (rule?.weight ?? 1) * SEVERITY_WEIGHT[rule?.severity ?? 'warning'];
    maxPenalty += max;
    if (status === 'fail') penalty += max;
    else if (status === 'warning') penalty += max / 2;
  }

  const score = maxPenalty === 0 ? 100 : Math.round((1 - penalty / maxPenalty) * 100);
  return {
    score: Math.max(0, Math.min(100, score)),
    passed,
    failed,
    warnings,
    notApplicable,
    bySeverity,
  };
}

/** Review one flow. Deterministic rules always run; the AI pass is opt-in. */
export async function reviewFlow(
  input: FlowReviewInput,
  ruleset: Ruleset,
  options: ReviewOptions = {},
): Promise<FlowReviewResult> {
  const flow = parseFlow(input.displayName, input.clientData);
  const findings: Finding[] = [];

  for (const rule of ruleset.rules) {
    if (!rule.enabled) continue;
    if (rule.evaluator === 'ai') continue; // handled in the AI pass
    const evaluator = evaluatorRegistry[rule.evaluator];
    if (!evaluator) {
      findings.push({
        ruleCode: rule.code,
        ruleName: rule.name,
        category: rule.category,
        status: 'not_applicable',
        severity: 'info',
        message: `No evaluator registered for "${rule.evaluator}".`,
      });
      continue;
    }
    findings.push(...evaluator({ flow, inventory: input.inventory, rule }));
  }

  if (options.ai && options.aiReviewer) {
    const aiFindings = await options.aiReviewer.review(flow, ruleset, findings);
    findings.push(...aiFindings.map((f) => ({ ...f, aiGenerated: true })));
  }

  return {
    displayName: input.displayName,
    flowId: input.inventory?.flowId,
    findings,
    score: scoreFindings(findings, ruleset.rules),
  };
}

File: plugin/Ember.Plugins/Engine/Reviewer.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace BetaRover.Ember.Engine
{
    /// <summary>Port of engine/src/reviewer.ts — deterministic review + scoring.
    /// The AI pass is intentionally omitted here; if wanted, the calling flow makes
    /// the Azure OpenAI HTTP call and appends its findings (keeps the plug-in free
    /// of outbound HTTP and secrets).</summary>
    public static class Reviewer
    {
        private static readonly Dictionary<string, int> SeverityWeight = new Dictionary<string, int>
        {
            { "info", 0 }, { "warning", 1 }, { "error", 3 }, { "critical", 6 }
        };

        private static readonly Dictionary<string, int> StatusOrder = new Dictionary<string, int>
        {
            { "not_applicable", 0 }, { "pass", 1 }, { "warning", 2 }, { "fail", 3 }
        };

        private static string WorstStatus(IEnumerable<Finding> findings)
        {
            string worst = "not_applicable";
            foreach (var f in findings)
                if (StatusOrder[f.Status] > StatusOrder[worst]) worst = f.Status;
            return worst;
        }

        public static ReviewScore ScoreFindings(List<Finding> findings, List<ReviewRule> rules)
        {
            var score = new ReviewScore();
            foreach (var f in findings)
            {
                switch (f.Status)
                {
                    case "pass": score.Passed++; break;
                    case "warning": score.Warnings++; score.BySeverity[f.Severity]++; break;
                    case "fail": score.Failed++; score.BySeverity[f.Severity]++; break;
                    case "not_applicable": score.NotApplicable++; break;
                }
            }

            var ruleByCode = new Dictionary<string, ReviewRule>();
            foreach (var r in rules) ruleByCode[r.Code] = r;

            var grouped = new Dictionary<string, List<Finding>>();
            foreach (var f in findings)
            {
                List<Finding> list;
                if (!grouped.TryGetValue(f.RuleCode, out list)) { list = new List<Finding>(); grouped[f.RuleCode] = list; }
                list.Add(f);
            }

            double penalty = 0, maxPenalty = 0;
            foreach (var kv in grouped)
            {
                string status = WorstStatus(kv.Value);
                if (status == "not_applicable") continue;
                ReviewRule rule; ruleByCode.TryGetValue(kv.Key, out rule);
                int weight = rule != null ? rule.Weight : 1;
                string sev = rule != null ? rule.Severity : "warning";
                double max = weight * SeverityWeight[sev];
                maxPenalty += max;
                if (status == "fail") penalty += max;
                else if (status == "warning") penalty += max / 2.0;
            }

            int computed = maxPenalty == 0 ? 100 : (int)Math.Round((1 - penalty / maxPenalty) * 100);
            score.Score = Math.Max(0, Math.Min(100, computed));
            return score;
        }

        /// <summary>Review one flow against a ruleset (deterministic rules only).</summary>
        public static FlowReviewResult ReviewFlow(string displayName, object clientData, FlowInventoryMeta inventory, Ruleset ruleset)
        {
            var flow = FlowParser.ParseFlow(displayName, clientData);
            var findings = new List<Finding>();

            foreach (var rule in ruleset.Rules)
            {
                if (!rule.Enabled) continue;
                if (rule.Evaluator == "ai") continue; // handled by the flow's AI pass, if any
                Func<EvaluatorContext, List<Finding>> evaluator;
                if (!Evaluators.Registry.TryGetValue(rule.Evaluator, out evaluator))
                {
                    findings.Add(new Finding
                    {
                        RuleCode = rule.Code, RuleName = rule.Name, Category = rule.Category,
                        Status = "not_applicable", Severity = "info",
                        Message = "No evaluator registered for \"" + rule.Evaluator + "\".",
                    });
                    continue;
                }
                findings.AddRange(evaluator(new EvaluatorContext { Flow = flow, Inventory = inventory, Rule = rule }));
            }

            return new FlowReviewResult
            {
                DisplayName = displayName,
                FlowId = inventory != null ? inventory.FlowId : null,
                Findings = findings,
                Score = ScoreFindings(findings, ruleset.Rules),
            };
        }
    }
}
